src/compute/src/logging/timely.rs

```diff
@@ -526,3 +327,403 @@ fn create_address_row(id: usize, worker: WorkerIdentifier, address: &[usize]) ->
+    requested_pow: Option<u128>,
+}
+
+impl Columnation for ParkDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct MessageDatum {
+    channel: usize,
+    worker: usize,
+}
+
+impl Columnation for MessageDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct ScheduleHistogramDatum {
+    operator: usize,
+    duration_pow: u128,
+}
+
+impl Columnation for ScheduleHistogramDatum {
+    type InnerRegion = CloneRegion<Self>;
+}
+
+/// Event handler of the demux operator.
+struct DemuxHandler<'a, 'b, 'c> {
+    /// State kept by the demux operator.
+    state: &'a mut DemuxState,
+    /// Demux output buffers.
+    output: &'a mut DemuxOutput<'b, 'c>,
+    /// The logging interval specifying the time granularity for the updates.
+    logging_interval_ms: u128,
+    /// The number of timely workers.
+    peers: usize,
+    /// The current event time.
+    time: Duration,
+    /// A capability usable for emitting outputs.
+    cap: &'a InputCapability<Timestamp>,
+}
+
+impl DemuxHandler<'_, '_, '_> {
+    /// Return the timestamp associated with the current event, based on the event time and the
+    /// logging interval.
+    fn ts(&self) -> Timestamp {
+        let time_ms = self.time.as_millis();
+        let interval = self.logging_interval_ms;
+        let rounded = (time_ms / interval + 1) * interval;
+        rounded.try_into().expect("must fit")
+    }
+
+    /// Handle the given timely event.
+    fn handle(&mut self, event: TimelyEvent) {
+        use TimelyEvent::*;
+
+        match event {
+            Operates(e) => self.handle_operates(e),
+            Channels(e) => self.handle_channels(e),
+            Shutdown(e) => self.handle_shutdown(e),
+            Park(e) => self.handle_park(e),
+            Messages(e) => self.handle_messages(e),
+            Schedule(e) => self.handle_schedule(e),
+            _ => (),
+        }
+    }
+
+    fn handle_operates(&mut self, event: OperatesEvent) {
+        let ts = self.ts();
+        let datum = OperateDatum {
+            id: event.id,
+            name: event.name.clone(),
+        };
+        self.output.operates.give(self.cap, (datum, ts, 1));
+
+        let datum = AddressDatum {
+            id: event.id,
+            address: event.addr.clone(),
+        };
+        self.output.addresses.give(self.cap, (datum, ts, 1));
+
+        self.state.operators.insert(event.id, event);
+    }
+
+    fn handle_channels(&mut self, event: ChannelsEvent) {
+        let ts = self.ts();
+        let datum = ChannelDatum {
+            id: event.id,
+            source: event.source,
+            target: event.target,
+        };
+        self.output.channels.give(self.cap, (datum, ts, 1));
+
+        let datum = AddressDatum {
+            id: event.id,
+            address: event.scope_addr.clone(),
```

Comment:
I think we should be able to avoid the `clone` by reordering this with the next statement.

Reply:
But then we'd have to `clone` the whole event when pushing it to `self.state.dataflow_channels`, so that wouldn't help us.